grammar-gen/src/codegen/rt.rs

```diff
@@ -0,0 +1,402 @@
+            Self_,
+            SelfUpper,
+            Static,
+            Struct,
+            Super,
+            Trait,
+            True,
+            Type,
+            Union,
+            Unsafe,
+            Use,
+            Where,
+            While,
+            Yield,
+
+            Abstract,
+            Become,
+            Box,
+            Do,
+            Final,
+            Macro,
+            Override,
+            Priv,
+            Try,
+            Typeof,
+            Unsized,
+            Virtual,
+
+            Literal,
+            Plus,
+            Minus,
+            Star,
+            Slash,
+            Percent,
+            Caret,
+            Not,
+            And,
+            Or,
+            AndAnd,
+            OrOr,
+            Shl,
+            Shr,
+            PlusEquals,
+            MinusEquals,
+            StarEquals,
+            SlashEquals,
+            PercentEquals,
+            CaretEquals,
+            AndEquals,
+            OrEquals,
+            ShlEquals,
+            ShrEquals,
+            Equals,
+            EqualsEquals,
+            NotEquals,
+            GreaterThan,
+            LessThan,
+            GreaterThanEquals,
+            LessThanEquals,
+            At,
+            Underscore,
+            Dot,
+            DotDot,
+            DotDotDot,
+            DotDotEquals,
+            Comma,
+            Semicolon,
+            Colon,
+            ColonColon,
+            RightArrow,
+            FatArrow,
+            Pound,
+            Dollar,
+            QuestionMark,
+
+            LParen,
+            RParen,
+            LBracket,
+            RBracket,
+            LBrace,
+            RBrace,
+
+            FragmentBlock,
+            FragmentExpr,
+            FragmentIdent,
+            FragmentItem,
+            FragmentLifetime,
+            FragmentLiteral,
+            FragmentMeta,
+            FragmentPat,
+            FragmentPath,
+            FragmentStmt,
+            FragmentTT,
+            FragmentTy,
+            FragmentVis,
+        }
+
+        use TokenDescription::*;
+
+        #[derive(Clone, Debug, Eq, Hash, PartialEq)]
+        pub struct RustParser {
+            buffer: TokenBuffer,
+            stack: Vec<State>,
+        }
+
+        #[derive(Clone, Debug, Eq, Hash, PartialEq)]
+        pub struct TransitionData {
+            pub popped: usize,
+            pub buf_size: usize,
+            // TODO: wrap this in an opaque type.
+            pub pushed: Vec<&'static str>,
+        }
+
+        impl RustParser {
+            #(
+                pub fn #entry_points() -> RustParser {
+                    RustParser {
+                        buffer: TokenBuffer::Empty([]),
+                        stack: vec![(#entry_points, concat!("<", stringify!(#entry_points), " entry point>"))]
+                    }
+                }
+            )*
+
+            pub fn step(&mut self, token: TokenDescription) -> Option<TransitionData> {
```

Comment:
That kind of error reporting is bad for error spanning. The parser has a lookahead of three. That means it is lagging three tokens behind any library that use it. We should not let it expect that the error happened at the position at which `token` is.

Instead, we should make `RustParser` generic over a `Span` type as we do in `expandable_impl`, require an additional `span: Span` parameter to the step function, and return that span if the parsing failed.